### ISSUE TITLE
Cyrillic fonts for InkStitch? (I'm willing to contribute)

### ISSUE THREAD

**Issue by toastrix** [inkstitch/inkstitch#861]
I'd like to see InkStitch support for the Cyrillic alphabet, which is used to write Russian, Ukrainian, Belarussian, and other Eastern European languages.

I'm an experienced InkStitch user and a programmer (and I speak Russian). I'd be happy to provide an SVG file with satin-stitch layouts for the characters (with underay, stitching beginning and ending at a consistent height, etc.) and Unicode codepoints for the letters. I've looked at the SVG files for the existing fonts, and don't think I can just send you an SVG file and expect it to work with a different alphabet, but I'd be happy to put together whatever would be needed.

I'm picturing this working like the other fonts - you'd go to the Lettering dialogue, choose the Cyrillic font, and type in Cyrillic, and end with an Ink/Stitch lettering object like the ones created for existing fonts.

If this is something the InkStitch team is interested in, please don't hesitate to get in touch. I think I'd need to know more about how InkStitch fonts are implemented in order to start working, so I'd like to get in touch with someone who knows the internals before I start on the SVG. Thank you for everything you do!

**Comment by Augusa** [comment]
I started last year with creating a font for Inkstitch. You can see the beginning of the result in DéjàVu font from inkstitch lettering. Now this font is finished and I hope it will be completed in the next release of Inkstitch. 
You can read several posts of my search in Font Editor for Ink/Stitch #586.
 We are now several working on this subject and you can read Lettering tutorial test + Kerning question #809
I am not at all an expert of these questions but I can explain you how we are working now to prepare a file which works in Inkstitch. 
I am french and it's a little difficult to explain in english but we can try if you don't find wath you search.

**Comment by toastrix** [comment]
I've read over the posts - thank you!

I read French, though I don't write it very well. So if at some point you'd prefer to explain in French, that's fine, too.

**Comment by LaureFR38** [comment]
@toastrix I am part of the French team of @Augusa.
I'm trying to improve the kerning of the fonts to be automatic. Could we help each other? I would need a little help with the python code as I am not very good at it. I could also explain to you. If this interests you send me an email at aliaslaure@gmail.com.

**Comment by toastrix** [comment]
I am slowly working through the needed characters and testing embroidery. I'll put up a 1.0 file when it's done.

I've read through the documentation on stroke fonts for the Hershey font that inspired InkStitch's built-in Geneva Simple Sans, and while I didn't want to copy from Geneva directly, I'm also using the Cyrillic characters from the Geneva font (not the built-in one, obviously!) as a reference to help make sure my characters have a similar style.

**Comment by Augusa** [comment]
It is also possible to create a font for Inkstitch from an existing font. You just need to choose a font with a licence compatible with Inkstitch like google fonts or font from the public domain and then you have to turn each glyph to embroidery. So you don't need to create the letters just to turn them to embroidery. It is already enough work.

**Comment by LaureFR38** [comment]
@toastrix 
Are you going to fill in the character spaces in the json? We are in the process of modifying the behavior of Inkstich for reading this json. Your work may be lost. Look here if you want to know more: https://github.com/inkstitch/inkstitch/pull/848

**Comment by toastrix** [comment]
@LaureFR38 - I hadn't made any decisions about the JSON yet; once I had all the lower-case characters done, I was going to copy the values from the Geneva-simple-sans .json and see how well they worked when the text was sewn.

**Comment by toastrix** [comment]
I've done all the upper- and lower-case glyphs for the Cyrillic letters used for modern Russian, Belarussian, Ukrainian, Serbian, Montenegrin, and Macedonian. They're satin-stitch, done in the same style as the geneva_simple font included with InkStitch.

I believe the next thing to do is create the SVG font files. I know how this is done for regular SVG fonts (one layer per character, named for the character, etc.) but I have some questions about the four files InkStitch uses for each font. 

The InkStitch file names use arrow characters (→, etc.), but when I compared running the InkStitch simulator on sample letters from different files, I didn't see a difference between them. What do I need to know to create the files in the correct format?

(I've read through existing font-related issues and the online documentation. If there's an existing guide that I've missed, please feel free to point me to it.)

Once I have the font files, I'll share them along with some pictures of embroidered sample text.

Also, the Montenegrin З́ isn't a single Unicode character - it's a composed character created from  З+◌́ (and doesn't seem to display correctly as I'm typing this comment!) Is it still possible to use it as a single glyph? If a user is typing in Montenegrin, cutting and pasting Montenegrin text, etc., they should be sending the right Unicode sequence. I'm just curious whether there are any issues caused by compound characters like this one.

Thank you all for your help - I appreciate it!

**Comment by Augusa** [comment]
The actual fonts of Inkstitch haven't  differences in the 4 files. When I sent DejaVu font with only one direction I was asked for the other one. But it was not yet done. Now I have created the 4 files because they prefer to have them when Inkstitch will be ready to use them. On my point of view it is much more important to have several different fonts. There is no kerning except for the upper case of DejaVu. But it will be done as you know for some new fonts. I don't know anything about montenegrin character but may be a long runnig stitch between the two parts will help. I have had some difficulties with a character for german not existing on my keyboard on the beginning. Now it works ... but I don't know why!

**Comment by lexelby** [comment]
Oh wow, this is so exciting!  @toastrix I'm really happy you're here and working on this.  I had been hoping someone might take on a Cyrillic font!

I'll try to answer your questions here.  First, I want to let you know that the Lettering extension is very much a work in progress, as @LaureFR38 mentioned.  I wrote it initially without being an expert in either fonts or unicode, and I'm afraid it shows.  The code is still very much in development, and I have no qualms about people changing it to work better.  If you would be interested in diving in, I would be happy to point you in the right direction and answer any questions you might have.  My time for development work has been limited lately, so I'm especially happy to hear that you're a programmer too!

Now, for your questions:

>  I've looked at the SVG files for the existing fonts, and don't think I can just send you an SVG file and expect it to work with a different alphabet,

That's actually exactly what I intended.  As it's written now, it really doesn't care what unicode characters you use.  It just goes through the text that the user inputs character by character.  For each character, it finds a glyph layer named with that character and uses the objects in that layer.  That should work for Cyrillic just the same as English or French.  I hope!

> Also, the Montenegrin З́ isn't a single Unicode character - it's a composed character created from З+◌́

...and here's where we run into my lack of Unicode experience.  I knew about composed characters, but I totally forgot about them when I wrote the Lettering code.  Hopefully, we can make the code pay attention to each character -- whether composed of multiple codepoints or not.  I have a suspicion the code might not handle this correctly right now.

> The InkStitch file names use arrow characters (→, etc.), but when I compared running the InkStitch simulator on sample letters from different files, I didn't see a difference between them. What do I need to know to create the files in the correct format?

The idea is that each file is for when text is being sown in the direction shown by the arrow.  Left-to-right, right-to-left, etc.  When sewing a few lines of text, for example, it's advantageous to sew the first line left-to-right, the second line right-to-left, and so on.  The problem is that the order that satins are sewn in needs to depend on what direction you're sewing the text in.

For example, consider a lower-case letter "t".  Ideally, the vertical part should sew first, and the cross-bar should sew on top of it, so that it looks like a writer has crossed the "t".

The Lettering extension uses the Auto-Route Satin Columns extension under the hood.  We handle this ordering issue by using the "preserve order" checkbox.

This solution is really quite far from ideal.  I wish we could make it so that you could just tell Ink/Stitch "sew the cross-bar last".  It's not easy, because often one would want to sew the cross bar after multiple satins, like for an "H".  And for a "#", we want to sew _both_ crossbars after the vertical portions.

We would ideally have some way of saying "sew these two objects before those two".  I think the code for that feature in the Auto-Route Satin extension would be manageable, but I haven't yet figured out how we would have the user express that concept to Ink/Stitch.  Maybe with SVG groups?

Anyway, bottom line is: if you want to see what the different arrow SVGs do, tell Lettering to sew the text in another direction.

> I believe the next thing to do is create the SVG font files. I know how this is done for regular SVG fonts (one layer per character, named for the character, etc.) 

BTW, I think you may already have figured this out, but just in case: Inkscape's built-in Typography extensions make this easy.  I use "Add Glyph Layer..." to make each new layer.  We _don't_ use the "Convert Glyph Layers to SVG Font" option though, since Ink/Stitch font SVGs are not actual SVG fonts.

**Comment by toastrix** [comment]
My InkStitch work will be interrupted for a few days due to hardware issues*, but thank you all for your help so far! Looking over my earlier posts, I can see where I was puzzling out how InkStitch handled fonts - I have a better handle on it now.

@Augusa, @lexelby - that was exactly what I needed to know about the four font files! I'll take a closer look at them - this is going to be an issue with some of the Cyrillic characters, too.

Let me fill in a little background about Cyrillic, and how I'm approaching it, for anyone who's interested in what I'm doing but isn't already familiar with the writing system.

Like the Latin alphabet that English uses, Cyrillic has upper- and lower-case letters, aligned on a baseline, with occasional ascenders and descenders. So while we might run into a surprising hiccup here and there, adding Cyrillic probably doesn't require the Lettering code to do anything all that different. (Arabic, with its connected, cursive script where letters have different shapes based on what's before and after them, would be another story!)

Of all the languages written in Cyrillic, Russian has far and away the most speakers - but it's also used to write other languages in Eastern Europe and the former Soviet republics. Many of them add letters that aren't used in Russian.

When I started doing the Cyrillic glyphs, I thought that instead of stopping with the 33 letters for Russian (66 glyphs, counting upper and lower case) I'd also do the fifteen or so more that covered the rest of the Slavic languages that use Cyrillic. (Supporting the non-Slavic languages would require a much larger number of glyphs.)

All of the Cyrillic characters in the most widely-spoken languages correspond to single Unicode codepoints. However, some of characters in languages like Montenegrin don't. Computers treat them as a combination of two codepoints, one for the letter form and one for the diacritic. This can potentially cause bugs in code that assumes one letter = one codepoint.

Because of the aforementioned hardware problems, it'll be a few days before I can test my font with my embroidery machine. If the composed characters work, so much the better. If they don't, I'll document them and provide whatever information would be needed to revisit the Lettering code, and save the composed characters for a later release.

(and my Brother machine's built-in Cyrillic font doesn't include those letters either, so either way my InkStitch font will support at least as many languages as Brother, haha)

As for creating the glyphs ... Some Cyrillic characters, especially the upper-case letters, have the same shapes as letters in the Latin alphabet. Cyrillic Н and English H have different Unicode codepoints, but since they look identical, I can use the existing Geneva Simple Sans satin stitches. Letters like Б don't have exact equivalents, but by ungrouping the satins from the Latin font, I can re-use the vertical strokes, curves, etc., and create Cyrillic letters with the right stroke widths and proportions to look like they belong with the rest of the font. And when I don't have everything I need, I'm creating new satins, comparing them to similar Geneva fonts to make sure I don't add anything too avant-garde.

It's been a learning experience, but I have my glyphs and I know what I need to make the four font files. Hopefully I'll have good news to report when my hardware issues are resolved!

*my cat threw up on my laptop keyboard

**Comment by lexelby** [comment]
My deepest sympathies for your keyboard.  I think I hear my cats vomiting in solidarity off in the distance...

Thank you, this is really useful information, and learning about Cyrillic is really interesting!  Since early on, I've wanted to make Ink/Stitch useful and inviting to non-English speakers, and getting the lettering right for many languages is table stakes, in my book.  Thank you for your work.

By the way, if you're interested in working on the Russian translation for Ink/Stitch, we can hook you up with that as well.  I know, one thing at a time. :)  I get ahead of myself sometimes.

**Comment by LaureFR38** [comment]
@lexelby @toastrix 
All languages ​​are interested in associating characters that do not have a unicode match. In some font for aesthetic reasons, some character pairs have been redesigned. For example look at "cx".

![image](https://user-images.githubusercontent.com/63351363/98459155-53d2c680-2198-11eb-81e3-f994a8b0f288.png)

I don't know what you think about it, but for me it is essential to go through fontforge to create the glyph layers in Inkscape in order to handle the kerning correctly otherwise all the work will have to be redone and it's a shame :-(

**Comment by Augusa** [comment]
The problem of differents glyphs for differents languages is endless. Some latin fonts have thousands of glyphs. You have necessary to choose for embroidery. I work also with Stitch Era. Very few fonts come with something else than the basics glyphs from english. The fonts for polish, czech, islandic or turckish... have a lot of glyphs different from french or german for exemple. In the last fonts on which  i am working now, I add glyphs for german but of course it is not complete for other languages. 
The solution on Stitch Era is the possibility to get embroidery from TTF fonts automatically. The result is never perfect but with some corrections it works. May be later Inkstitch will have this fonctionnality. But now it's already very well if we are able to add several fonts even if not with thousands of glyphs and I agree with @LaureFR38 that the kerning is important. I started to do it with DejaVu manually it is simply mad. 
@lexelby I would be happy to know if it is possible to have only the file from left to right for fonts as Lobster in which lettres have to cover a small part of the end of the letter to look unbroken. In this case the backward possibility produce great troubles because the small parts are on the top and not under the following letter in the order of writing. Is it possible not to allow the backward for a particuliar font?

**Comment by lexelby** [comment]
Yes!  Just include the one file, `→.svg`.  Ink/Stitch will use that for all directions.  It will always use `→.svg` as a fallback if the other directions don't exist.

**Comment by Augusa** [comment]
![almost1](https://user-images.githubusercontent.com/54963659/98467217-6b2fa500-21d4-11eb-8eeb-f2be3945aac4.jpg)
![almost2](https://user-images.githubusercontent.com/54963659/98467219-6d91ff00-21d4-11eb-9afa-9b3ca8d7571f.jpg)
I try to explain the difficulty with picture. The backward is not usable but the worst is that it is not easy to see on a simulation. If somebody enter a text with 2 lines or more he probably will see the bad result only after having embroidered it. How to prevent an user to use backward? Hope this is easier to understand as my bad english

**Comment by lexelby** [comment]
Oh!  You want to remove the backward option entirely for your font.  Yes, that's a great idea.  Could you please open a new issue, labeled with "feature request", asking for this?  I don't want to forget this idea.

**Comment by toastrix** [comment]
Now that I have the →.svg and ←.svg files, what do I need to do to get my font to show up in the Lettering dialog box?

I created a folder for my font in Inkstitch's fonts folder and added the .json file and the two font files. I also experimented with creating a .inkstitch folder in my %HOMEPATH% directory with a fonts subdirectory and adding the files there, since it looks as though the Lettering code looks there, too. (In my experience, though, Windows 10 software is inconsistent in how it handles the idea of a home directory, and it's not unusual to find it using something other than the value of the HOMEPATH environment variable.)

I was also wondering how InkStitch determines where the baseline is. It looks as though the included fonts don't specify global settings for cap height, descent, etc. I used the named guides from the Typography Canvas to set up my font file - is that enough, or is there something else I need to do? Thanks!

**Comment by Augusa** [comment]
I suggest to create 4 files in your folder. You just copy one existing for the 2 missing. It was right for me.

**Comment by lexelby** [comment]
Strange, I'm not sure why your font is not being recognized.  Are you using a downloaded version of Ink/Stitch, or did you set up a development environment?

Ink/Stitch does use the baseline guide from the Typography canvas.  It ignores the other guides (at least, it doesn't use them yet...).

**Comment by lexelby** [comment]
Closing for now, but feel free to reopen if if you decide to resume work on this.

**Comment by kaalleen** [comment]
Adding cyrillic fonts should be no issue on the programs side. Would be nice to have contributions :)
I'll close the issue though, since Ink/Stitch is ready to process the fonts. If there is anything that needs to be done or a font to contribute, please open a new issue.
